src/main/java/org/jenkinsci/plugins/workflow/support/pickles/serialization/RiverReader.java

```diff
@@ -151,7 +153,7 @@ public ListenableFuture<Unmarshaller> restorePickles(Collection<ListenableFuture
         config.setClassResolver(new SimpleClassResolver(classLoader));
         //config.setSerializabilityChecker(new SerializabilityCheckerImpl());
         config.setObjectResolver(combine(evr, ownerResolver));
-        Unmarshaller eu = new RiverMarshallerFactory().createUnmarshaller(config);
+        Unmarshaller eu = new TracingRiverMarshallerFactory().createUnmarshaller(config);
```

Comment:
The overridden methods will increase the stack depth by 1 for each nested object, so if issues like [JENKINS-52966](https://issues.jenkins.io/browse/JENKINS-52966) are not due to cycles but just deeply nested objects because of the way Declarative is designed (which seems quite possible given user comments in that issue), this will make the problem worse. Could we somehow use the old code path by default, but then whenever we hit a `StackOverflowError` re-deserialize things with `TracingRiverMarshallerFactory` enabled? Or maybe only enable tracing if a system property is set?

Reply:
Exactly. If this PR does turn out to be capable of diagnosing an issue in this area, we would need to decide how to activate it.